impl/src/main/java/org/jboss/weld/bootstrap/events/BuilderInterceptorBean.java

```diff
@@ -0,0 +1,168 @@
+package org.jboss.weld.bootstrap.events;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import javax.enterprise.context.Dependent;
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.enterprise.inject.spi.InterceptionType;
+import javax.enterprise.inject.spi.Interceptor;
+import javax.enterprise.inject.spi.Prioritized;
+import javax.interceptor.InvocationContext;
+
+import org.jboss.weld.context.WeldCreationalContext;
+import org.jboss.weld.logging.BeanLogger;
+
+/**
+ * @author Tomas Remes
+ */
+public class BuilderInterceptorBean<T> implements Interceptor<T>, Prioritized {
+
+    private final Set<Annotation> bindings;
+
+    private final int priority;
+
+    private final InterceptionType interceptionType;
+
+    private final Function<InvocationContext, Object> interceptorFunction;
+
+    private final BiFunction<InvocationContext, Bean<?>, Object> interceptorMetadataFunction;
+
+    public BuilderInterceptorBean(Set<Annotation> interceptorBindings, InterceptionType type, int priority,
+            Function<InvocationContext, Object> interceptorFunction) {
+        this.bindings = interceptorBindings;
+        this.priority = priority;
+        this.interceptionType = type;
+        this.interceptorFunction = interceptorFunction;
+        this.interceptorMetadataFunction = null;
+    }
+
+    public BuilderInterceptorBean(Set<Annotation> interceptorBindings, InterceptionType type, int priority,
+            BiFunction<InvocationContext, Bean<?>, Object> interceptorFunction) {
+        this.bindings = interceptorBindings;
+        this.priority = priority;
+        this.interceptionType = type;
+        this.interceptorMetadataFunction = interceptorFunction;
+        this.interceptorFunction = null;
```

Comment:
It might be better to have a private constructor taking `interceptorMetadataFunction` and `interceptorFunction` as params and two convenient public constructors for each type of function.